### PR TITLE
Fixes #22468: Fix escaping in rudderc

### PR DESCRIPTION
**Pull request by amousset**
https://issues.rudder.io/issues/22468

### CODE REVIEW

**Comment by amousset** [comment]
PR updated with a new commit

**Comment by amousset** [comment]
PR updated with a new commit

File: policies/rudderc/src/backends/unix/cfengine.rs [hunk 1]
```diff
@@ -43,6 +45,118 @@ pub fn expanded(s: &str) -> String {
...
+        })
+        .collect::<Vec<u8>>();
+    std::str::from_utf8(&s)
+        .unwrap_or_else(|_| panic!("Canonify failed on {}", input))
+        .to_owned()
+}
+
+/// Canonify a condition before using it in `if`/`unless` attributes.
+///
+/// We need to take special care to preserver variable expansion as
```
Comment (Fdall) [inline policies/rudderc/src/backends/unix/cfengine.rs:118]:
```suggestion
/// We need to take special care to preserve variable expansion as
```

File: policies/rudderc/src/backends/unix/cfengine.rs [hunk 2]
```diff
@@ -43,6 +45,118 @@ pub fn expanded(s: &str) -> String {
...
+    fn it_cfengine_canonifies() {
+        assert_eq!(cfengine_canonify(""), "".to_string());
+        assert_eq!(cfengine_canonify("abc"), "abc".to_string());
+        assert_eq!(cfengine_canonify("a-bc"), "a_bc".to_string());
+        assert_eq!(cfengine_canonify("a_bc"), "a_bc".to_string());
+        assert_eq!(cfengine_canonify("a bc"), "a_bc".to_string());
+        assert_eq!(cfengine_canonify("aàbc"), "a__bc".to_string());
+        assert_eq!(cfengine_canonify("a&bc"), "a_bc".to_string());
+        assert_eq!(cfengine_canonify("a9bc"), "a9bc".to_string());
+        assert_eq!(cfengine_canonify("a😋bc"), "a____bc".to_string());
```
Comment (Fdall) [inline policies/rudderc/src/backends/unix/cfengine.rs:160]: I feel like a test using a variable expression is needed here

File: policies/rudderc/src/backends/unix/cfengine.rs [hunk 3]
```diff
@@ -43,6 +45,118 @@ pub fn expanded(s: &str) -> String {
...
+}
+
+/// Canonify a condition before using it in `if`/`unless` attributes.
+///
+/// We need to take special care to preserver variable expansion as
+/// our classes include the class_parameter variable value.
+///
+/// To do so, we split the condition to only canonify text
+/// outside of variables.
+fn cfengine_canonify_condition(c: &str) -> String {
```
Comment (Fdall) [inline policies/rudderc/src/backends/unix/cfengine.rs:123]: Could you write some simple tests for this?

**Comment by amousset** [comment]
PR updated with a new commit

**Comment by amousset** [comment]
PR updated with a new commit

File: policies/rudderc/src/ir/condition.rs [hunk 4]
```diff
@@ -73,7 +98,7 @@ impl FromStr for Condition {
         fn valid_char(c: char) -> bool {
             // Ideally, we could parse and validate condition expressions.
             // For now, let's only check for disallowed chars instead.
-            let valid_chars = ['"', '$', '{', '}', '|', '&', ' '];
+            let valid_chars = ['$', '{', '}', '|', '&', '.', ' ', '!', '_'];
```
[the remark this answers is not in the document]
Comment (amousset) [reply policies/rudderc/src/ir/condition.rs:101]: nice catch, a mistake

**Comment by amousset** [comment]
PR updated with a new commit

**Comment by amousset** [comment]
@Fdall updated

**Review by Fdall — APPROVED** [review]

**Comment by amousset** [comment]
OK, squash merging this PR

**Pull request merged**